### PR TITLE
drivers: led_strip: modernize ws2812 drivers

### PR DESCRIPTION
**Pull request by mbolivar**
This is the only led_strip driver using GPIO.

@jhedberg are you still using this, btw?

### CODE REVIEW

**Comment by mbolivar** [comment]
@mnkp this driver is exactly why I would like to have a "fast" gpio API, for reference.

**Review by pabigot — COMMENTED** [review]
This looks trivial, but I believe the desire is to take advantage of the fact we're touching these drivers to also complete the conversion to devicetree at the same time.

Looking at this my immediate questions are "what's the active level" (which would normally be or'd in from the devicetree node), and "should this be `GPIO_OUTPUT_INIT_LOW` or `_HIGH`".

**Comment by mbolivar** [comment]
> This looks trivial, but I believe the desire is to take advantage of the fact we're touching these drivers to also complete the conversion to devicetree at the same time.

If I'm going to do that, I want to know if people think I should create a unified set of bindings for this and the other driver for the same LED strip.

Apart from the GPIO driver there is also a SPI driver for the same hardware. I was thinking of unifying the bindings similarly to how devices with I2C or SPI based drivers work (e.g. https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/iio/accel/lis302.txt).

Worth doing?

**Comment by pabigot** [comment]
> Worth doing?

I couldn't say.  The bitbang implementation seems significantly more limited than the SPI one, e.g. wrt pixel order.  If the bindings are unified, should we try to support common functionality, or just blow up if the devicetree properties don't match the implemented behavior?  Is there value in evolving the bitbang driver or is it obsolete so we should deprecate it and let it expire unmolested (beyond the minimal necessary changes currently in this PR)?

If it's worth enhancing it then a unified binding seems to make sense.  E.g. the way lis2dh (or eeprom in #19972) is done where the interface yaml includes common yaml.

**Comment by jhedberg** [comment]
> @jhedberg are you still using this, btw?

@mbolivar not actively, but I do have the HW. It was used for some micro:bit demos: https://www.kitronik.co.uk/5625-zip-halo-for-the-bbc-microbit.html

**Comment by mbolivar** [comment]
>  Is there value in evolving the bitbang driver or is it obsolete so we should deprecate it and let it expire unmolested (beyond the minimal necessary changes currently in this PR)?

I ultimately want the bit-banging driver to be the "main" driver, but I can't do that portably without fast generic GPIO routines, which of course we have deferred work on. (The SPI driver's use of memory is an obscenity, but it's the only portable way to manage this thing that I could find within Zephyr's current driver APIs.)

So I guess what I'll do is create a separate binding for the SPI driver, and let the bit-bang driver take the main `worldsemi,ws2812` compatible. That will be a breaking change, but I don't think anybody's using this stuff. I just wrote it for fun a while ago.

> @mbolivar not actively, but I do have the HW. It was used for some micro:bit demos: https://www.kitronik.co.uk/5625-zip-halo-for-the-bbc-microbit.html

OK, let's keep it then, and generalize it when we have the GPIO driver API to make it possible to get rid of the inline assembly.

**Comment by sslupsky** [comment]
Stumbled across this issue today.  You may want to have a look at this repo if you are not already familiar with it.  Looks like a pretty well used and designed led driver api.

https://github.com/FastLED/FastLED

**Comment by mnkp** [comment]
This driver would be a good testcase for fast GPIO API (yes, we need it).

The PR could be approved as is, but @mbolivar if you're planning to add DTS bindings we're going to wait :-)

**Review by jhedberg — APPROVED** [review]

**Comment by mbolivar** [comment]
> The PR could be approved as is, but @mbolivar if you're planning to add DTS bindings we're going to wait :-)

I'm halfway done with this. I have the SPI driver "fully" dt-ized (right now the conversion was only partial) and still need to do the DT one.

I'll try to get this done tomorrow.

**Comment by mbolivar** [comment]
> Stumbled across this issue today. You may want to have a look at this repo if you are not already familiar with it. Looks like a pretty well used and designed led driver api.
> 
> https://github.com/FastLED/FastLED

Thanks for the pointer.

We are familiar with that library -- you can see the discussion in https://github.com/zephyrproject-rtos/zephyr/issues/11917 for details, for example it is mentioned here:

https://github.com/zephyrproject-rtos/zephyr/issues/11917#issuecomment-445063860

FastLED makes extensive use of C++ templates to get its job done, though, so we need a different approach for Zephyr.

**Comment by sslupsky** [comment]
Thank you @mbolivar for the link.  Interesting read indeed.

I do not quite understand the meaning of your last comment regarding C++.  Is Zephyr not intended to be used in a C++ environment?

**Comment by pabigot** [comment]
By [policy](https://docs.zephyrproject.org/latest/reference/kernel/other/cxx_support.html) Zephyr currently disallows use of C++ except within applications, so it's not suitable for a raw GPIO interface.

**Comment by mbolivar** [comment]
Right -- we can only use C (and assembly) from drivers, which is the layer we're discussing here.

**Comment by jfischer-phytec-iot** [comment]
> Right -- we can only use C (and assembly) from drivers, which is the layer we're discussing here.

I would be happy if this driver continues to use assembler, it is a very nice example.

**Comment by mbolivar** [comment]
> I would be happy if this driver continues to use assembler, it is a very nice example.

I would be unhappy if the driver continues to use assembler :).

It is not portable. This implementation is nrf51 only.

**Comment by jfischer-phytec-iot** [comment]
> > I would be happy if this driver continues to use assembler, it is a very nice example.
> 
> I would be unhappy if the driver continues to use assembler :).
> 
> It is not portable. This implementation is nrf51 only.

a) anyway a good example b) can be made portable, at least for Cortex-Mx

**Comment by mbolivar** [comment]
> > > I would be happy if this driver continues to use assembler, it is a very nice example.
> > 
> > 
> > I would be unhappy if the driver continues to use assembler :).
> > It is not portable. This implementation is nrf51 only.
> 
> a) anyway a good example

drivers/ is not a place for example code.

> b) can be made portable, at least for Cortex-Mx

The point of https://github.com/zephyrproject-rtos/zephyr/issues/11917 is to make the GPIO portions portable for all devices supported by zephyr, not just some.

The nop-loops in the assembly are not GPIO related of course, but I see the precision timeout work that was recently merged as a good starting point for making that portable too, also without assembly.

**Comment by jfischer-phytec-iot** [comment]
> The nop-loops in the assembly are not GPIO related of course, but I see the precision timeout work that was recently merged as a good starting point for making that portable too, also without assembly.

I think it would not help, you need to lock interrupts anyway to fulfill the timings.

**Comment by mbolivar** [comment]
> > The nop-loops in the assembly are not GPIO related of course, but I see the precision timeout work that was recently merged as a good starting point for making that portable too, also without assembly.
> 
> I think it would not help, you need to lock interrupts anyway to fulfill the timings.

Of course you need to call irq_lock() while you are sending the pulses.

My point is that now we have a portable way to convert nanoseconds to cycles, e.g. `k_ns_to_cyc_ceil32()`. That feels like a necessary ingredient to know how to portably halt in while generating the pulse. Am I missing something here?

**Comment by jfischer-phytec-iot** [comment]
> My point is that now we have a portable way to convert nanoseconds to cycles, e.g. `k_ns_to_cyc_ceil32()`. That feels like a necessary ingredient to know how to portably halt while generating the pulse. Am I missing something here?

No, that is indeed great, I plan to use it in bit-bang swd driver but to determine the number of nops. But TBH I have no idea how it would be useful to use it with C code.

**Comment by mbolivar** [comment]
> I'm halfway done with this. I have the SPI driver "fully" dt-ized (right now the conversion was only partial) and still need to do the DT one

I meant GPIO one.

> I'll try to get this done tomorrow.

OK, I've DT-ized and new-gpio-api-ized everything, including the sample.

DNM for now because

- I need to retest the SPI driver on nRF52
- I need to add a bbc_microbit overlay

But should be good to review

File: samples/drivers/led_ws2812/boards/bbc_microbit.overlay [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	led_strip: ws2812 {
+		compatible = "worldsemi,ws2812-gpio";
+		label = "WS2812";
+
+		chain-length = <16>; /* arbitrary; change at will */
```
Comment (jhedberg) [inline samples/drivers/led_ws2812/boards/bbc_microbit.overlay]: Could we make this 24 since that's how many LEDs the Zip Halo has: https://www.kitronik.co.uk/5625-zip-halo-for-the-bbc-microbit.html

That would also match what `samples/drivers/led_ws2812/src/main.c` used to have with the _sw driver (now called gpio).

Comment (mbolivar) [reply samples/drivers/led_ws2812/boards/bbc_microbit.overlay]: Done, thanks. I like your new avatar btw.

**Review by dbkinder — APPROVED** [review]
doc changes look OK, thanks.

File: samples/drivers/led_ws2812/Kconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (mnkp) [inline samples/drivers/led_ws2812/Kconfig]: Is this file really required? Boards that want to use SPI driver enable it in their respective `.conf` files and for GPIO based bit-bang driver SPI shouldn't be enabled.

File: drivers/led_strip/ws2812_gpio.c [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+#define WS2812_PIN          DT_INST_0_WORLDSEMI_WS2812_GPIO_IN_GPIOS_PIN
+#define WS2812_PIN_FLAGS    DT_INST_0_WORLDSEMI_WS2812_GPIO_IN_GPIOS_FLAGS
+
+#define RED_ORDER    DT_INST_0_WORLDSEMI_WS2812_GPIO_RED_ORDER
+#define GREEN_ORDER  DT_INST_0_WORLDSEMI_WS2812_GPIO_GREEN_ORDER
+#define BLUE_ORDER   DT_INST_0_WORLDSEMI_WS2812_GPIO_BLUE_ORDER
+#define WHITE_ORDER  DT_INST_0_WORLDSEMI_WS2812_GPIO_WHITE_ORDER
+
+#if (DT_INST_0_WORLDSEMI_WS2812_SPI_HAS_WHITE_CHANNEL == 1)
+#define HAS_WHITE 1
```
Comment (mnkp) [inline drivers/led_strip/ws2812_gpio.c]: Couldn't it be written simply as
```
#define HAS_WHITE DT_INST_0_WORLDSEMI_WS2812_SPI_HAS_WHITE_CHANNEL
```
I guess the concern is that `DT_INST_0_WORLDSEMI_WS2812_SPI_HAS_WHITE_CHANNEL` may be undefined. But even if this is the case the `#if HAS_WHITE` will work. Otherwise this define is generated from `has-white-channel` property of type binary.

File: drivers/led_strip/ws2812_spi.c [hunk 4]
```diff
@@ -10,54 +11,68 @@
...
+#define SPI_OPER     (SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | \
+		      SPI_WORD_SET(8) | SPI_LINES_SINGLE)
+#define SPI_FREQ     DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_MAX_FREQUENCY
+#define ONE_FRAME    DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_ONE_FRAME
+#define ZERO_FRAME   DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_ZERO_FRAME
+
+#define RED_ORDER    DT_INST_0_WORLDSEMI_WS2812_SPI_RED_ORDER
+#define GREEN_ORDER  DT_INST_0_WORLDSEMI_WS2812_SPI_GREEN_ORDER
+#define BLUE_ORDER   DT_INST_0_WORLDSEMI_WS2812_SPI_BLUE_ORDER
+#define WHITE_ORDER  DT_INST_0_WORLDSEMI_WS2812_SPI_WHITE_ORDER
```
Comment (mnkp) [inline drivers/led_strip/ws2812_spi.c]: We probably should create `struct ws2812_config` and place all these data there. It will make it easier to extend the driver in the future by adding extra instances. The same could be done for `ws2812_gpio.c`.

**Review by mnkp — COMMENTED** [review]
LGTM with some minor comments.

File: samples/drivers/led_ws2812/Kconfig (hunk 2 above)
Comment (mbolivar) [reply samples/drivers/led_ws2812/Kconfig]: Why would boards want to have logic for this sample in their board files?

My guess is most people will wire an Adafruit ring to an existing dev board. I don't know of any boards with built-in rings.

File: drivers/led_strip/ws2812_gpio.c (hunk 3 above)
Comment (mbolivar) [reply drivers/led_strip/ws2812_gpio.c]: You're right, thanks.

File: drivers/led_strip/ws2812_spi.c (hunk 4 above)
Comment (mbolivar) [reply drivers/led_strip/ws2812_spi.c]: I'm trying to keep the loops operating on build time data rather than run-time data, so the compiler can optimize better.

I'd actually rather delete support for this whole feature than add flash and run-time overhead for it. I've seen only GRB or GRBW; I threw this in just because I heard there was some variation possible somewhere on the internet :).

What do you think?

**Comment by mbolivar** [comment]
@mnkp thanks! I've responded to your comments.

File: drivers/led_strip/ws2812_spi.c (hunk 4 above)
Comment (mnkp) [reply drivers/led_strip/ws2812_spi.c]: I believe I haven't explained myself very well. This is meant to be a fully fledged driver, even if we start support with one specific target. I'm concern about the situation where user wants to connect more than one WS2813 led strip to their board (having two or more SPI peripherals or using different set of GPIO pins in case of bit-bang driver). This is handled by creating multiple instances of the driver. In this case all the instance specific data need to go into the config struct available to the driver via `dev->config->config_info`. So my comment concerns all `DT_INST_0_` defines (I used multi-line comment but they don't stand out very well).

It's true that in this case the amount of optimizations the compiler can perform is smaller.

Comment (pabigot) [reply drivers/led_strip/ws2812_spi.c]: > This is meant to be a fully fledged driver, even if we start support with one specific target.

It would be nice if drivers generally supported multiple instances, but Zephyr is a long way from reaching that goal, and there are cases like this where it's simply not reasonable.  I'm open to a consensus decision that this driver be documented as supporting only a single device instance.

If it consensus is unreachable I would like to see this converted to an example that's modularized enough to allow it to be copied into applications that need it.  It doesn't sound like this particular implementation is suited to being a "fully fledged driver" even though it uses the API and is certainly valuable.

Comment (mbolivar) [reply drivers/led_strip/ws2812_spi.c]: > m concern about the situation where user wants to connect more than one WS2813 led strip to their board (having two or more SPI peripherals or using different set of GPIO pins in case of bit-bang driver). This is handled by creating multiple instances of the driver.

Yup, I understand what you're saying.

What I'm saying is that if we're going to make this driver multi-instance, I would rather delete the order configuration support since I think it's likely to be rarely needed, than inflate the flash and runtime requirements of the driver. Either way is fine with me.

**Comment by mbolivar** [comment]
kicking ci

File: drivers/led_strip/ws2812_spi.c (hunk 4 above)
Comment (mnkp) [reply drivers/led_strip/ws2812_spi.c]: > if we're going to make this driver multi-instance (even hypothetically in the future), I would rather delete the order configuration support since I think it's likely to be rarely needed

Please treat my comment regarding multiple instances as a suggestion. I'm fine with developing this driver in steps. That said, if we don't plan to support the order configuration long term it might be better to drop it now. I do believe that at some time point we should add support for multiple instances. On the other hand "order configuration" could be made a conditional code enabled via Kconfig option. That would also be a viable approach so no need to hurry here with the decisions either.

**Comment by mbolivar** [comment]
@mnkp please take another look at ws2812_gpio.c when you get a chance. I've made it multi-instance, but deleted the color channel order control. If that's OK for you I'll update the SPI driver in the same way and we should be good to go.

Thanks!

**Comment by mbolivar** [comment]
@jhedberg I've moved the assembly around a bit so it'll be easier to swap out when we get the "fast gpio" routines. In doing so, I noticed that this constraint also looked incorrect:

```
				[p] "r" (pin)
```

This should be "l", not "r", AFAICT: the Arm ARM I checked said both Rd and Rn need to be r0-r7 for the str instruction. Mostly an FYI, but please check if you want.

**Comment by jhedberg** [comment]
@mbolivar looks nice and clean! Are you verifying the correctness of the generated signal with actual HW, a logic analyser, or some other way? I'll try to find some time to test this PR on the ZIP Halo I have.

**Comment by mbolivar** [comment]
> @mbolivar looks nice and clean!

Thanks!

>  Are you verifying the correctness of the generated signal with actual HW, a logic analyser, or some other way? I'll try to find some time to test this PR on the ZIP Halo I have.

I've got a Saleae hooked up to an nrf51_pca10028 for GPIO driver timing.

- 0 bits are ~.35 usec wide
- 1 bits ~0.85 usec
- inter-bit times are 1.25-1.6 usec depending on the transition

The 1 bits are pushing the limit if you believe the datasheet, but are well within the [empirical limits](https://wp.josh.com/2014/05/13/ws2812-neopixels-are-not-so-finicky-once-you-get-to-know-them/) in the blog post. I could drop a nop from T1H if you'd like to make that tighter.

I am also visually testing on a 16 pixel GRBW ring from Adafruit.

**Comment by mbolivar** [comment]
Marked DNM until @mnkp has a chance to review, since I need to make SPI driver changes depending on his response.

**Comment by mnkp** [comment]
> @mbolivar looks nice and clean!

I second! Sorry for a delayed review, not much free time these days.

**Comment by mbolivar** [comment]
> Sorry for a delayed review, not much free time these days.

No problem! There was no rush, I just didn't want it merged halfway done.

I've updated the SPI driver to match the GPIO changes. I had to put back the Kconfig file in the led_ws2812 sample to make the build system select the SPI driver by default.

Leaving DNM on for now as I'm traveling for a US holiday and won't re-test the SPI driver until next week.

File: dts/bindings/led_strip/worldsemi,ws2812-gpio.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+  The GPIO driver uses inline assembly, and isn't available for all
+  boards.
+
+compatible: "worldsemi,ws2812-gpio"
+
+include: [base.yaml, ws2812.yaml]
+
+properties:
+  label:
```
Comment (mnkp) [inline dts/bindings/led_strip/worldsemi,ws2812-gpio.yaml]: This property should be likely moved to `dts/bindings/led_strip/ws2812.yaml` as it is not specific to ws2812-gpio. The `worldsemi,ws2812-spi.yaml` doesn't specify it. However, the relevant define is generated also for the spi version of the driver.

File: drivers/led_strip/Kconfig.ws2812 [hunk 6]
```diff
@@ -7,93 +9,33 @@
...
-	  If the green channel is shifted out first, specify 0.
-	  If second, specify 1, and so on.
-
-config WS2812_BLU_ORDER
-	int "Order in which a blue pixel should be shifted out"
-	default 2
-	range 0 3
+config WS2812_STRIP_SPI
+	bool "Enable the SPI driver"
+	depends on SPI
```
Comment (mnkp) [inline drivers/led_strip/Kconfig.ws2812]: We could change it to `select`, there would be no need for `samples/drivers/led_ws2812/Kconfig` in this case?

File: drivers/led_strip/ws2812_gpio.c [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+static int ws2812_gpio_update_rgb(struct device *dev, struct led_rgb *pixels,
+				  size_t num_pixels)
+{
+	const struct ws2812_gpio_cfg *config = dev->config->config_info;
+	const bool has_white = config->has_white;
+	u8_t *ptr = (u8_t *)pixels;
+	size_t i;
+
+	/* Convert from RGB to on-wire format (GRB or GRBW) */
+	for (i = 0; i < num_pixels; i++) {
```
Comment (mnkp) [inline drivers/led_strip/ws2812_gpio.c]: `for (size_t i = 0; i < num_pixels; i++) {` ?

**Review by mnkp — COMMENTED** [review]
LGTM with minor comments.

File: drivers/led_strip/Kconfig.ws2812 (hunk 6 above)
Comment (mnkp) [reply drivers/led_strip/Kconfig.ws2812]: Since we don't have `HAS_SPI` using `select` here is not such a great idea.

File: drivers/led_strip/Kconfig.ws2812 [hunk 8]
```diff
@@ -7,93 +9,33 @@
...
-comment "The following options determine channel data order on the wire."
-
-config WS2812_RED_ORDER
-	int "Order in which a red pixel should be shifted out"
-	default 1
-	range 0 3
-	help
-	  If the red channel is shifted out first, specify 0.
-	  If second, specify 1, and so on.
+choice WS2812_STRIP_DRIVER
```
Comment (mnkp) [inline drivers/led_strip/Kconfig.ws2812]: Since we allow for multiple instances maybe we should also allow to have a mix, e.g. one instance of GPIO and one instance of SPI based driver. In this case we shouldn't have a choice but two independent options.

File: drivers/led_strip/ws2812_spi.c [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+/*
+ * SPI master configuration:
+ *
+ * - mode 0 (the default), 8 bit, MSB first (arbitrary), one-line SPI
+ * - no shenanigans (don't hold CS, don't hold the device lock, this
+ *   isn't an EEPROM)
+ */
+#define SPI_OPER (SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | \
+		  SPI_WORD_SET(8) | SPI_LINES_SINGLE)
```
[the remark this answers is not in the document]
Comment (mbolivar) [reply drivers/led_strip/ws2812_spi.c]: I'll add a couple of macros for magic numbers, but the value 8 appears sometimes because a struct led_rgb color channel is defined as a u8_t, so I'm going to leave those (such as in ws2812_spi_ser()) as-is.

File: drivers/led_strip/ws2812_spi.c [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+ * Serialize an 8-bit color channel value into an equivalent sequence
+ * of SPI frames, MSbit first, where a one bit becomes SPI frame
+ * one_frame, and zero bit becomes zero_frame.
+ */
+static inline void ws2812_spi_ser(u8_t buf[8], u8_t color,
+				  const u8_t one_frame, const u8_t zero_frame)
+{
+	int i;
+
+	for (i = 0; i < 8; i++) {
```
[the remark this answers is not in the document]
Comment (mbolivar) [reply drivers/led_strip/ws2812_spi.c]: I hope you're OK with me leaving this as-is, since as you say it's subjective and the suggestion doesn't concern the GPIO API.

File: dts/bindings/led_strip/worldsemi,ws2812-gpio.yaml (hunk 5 above)
Comment (mbolivar) [reply dts/bindings/led_strip/worldsemi,ws2812-gpio.yaml]: The SPI yaml doesn't need this because spi-device.yaml (which it includes) already has this line.

Since the gpio binding inherits from base.yaml instead, it's needed here.

File: drivers/led_strip/Kconfig.ws2812 (hunk 6 above)
Comment (mbolivar) [reply drivers/led_strip/Kconfig.ws2812]: I think your second comment means I should disregard your first, which I will do. Please let me know if I'm wrong about that.

File: drivers/led_strip/Kconfig.ws2812 (hunk 8 above)
Comment (mbolivar) [reply drivers/led_strip/Kconfig.ws2812]: I think the changes I've made in this PR have already gotten quite a bit larger than the original request, which was to update it to the new GPIO API only, so I hope you're OK with me waiting to do this until someone actually needs it.

In my opinion, having multiple strips is already a bit of a stretch. Practically speaking, an application using this hardware is going to have one logical strip, since the device hardware is a daisy chain.

I can't think of a practical situation in which I would design an embedded device which would want both drivers enabled in software.

File: drivers/led_strip/ws2812_gpio.c (hunk 7 above)
Comment (mbolivar) [reply drivers/led_strip/ws2812_gpio.c]: I prefer to keep this in the traditional Linux kernel style.

**Comment by mbolivar** [comment]
I've retested on nrf51_pca10028 and nrf52_pca10040, so removing DNM.

@mnkp I've responded to your comments, please take another look when you get a chance.

File: drivers/led_strip/Kconfig.ws2812 (hunk 8 above)
Comment (mnkp) [reply drivers/led_strip/Kconfig.ws2812]: Sure, that was a major rework and the code looked good from the very beginning.

**Review by mnkp — APPROVED** [review]
